File: packages/snow_draw_core/lib/draw/edit/preview/edit_preview_engine.dart
import '../../models/draw_state.dart';
import '../../models/interaction_state.dart';
import '../core/edit_errors.dart';
import '../edit_operation_registry_interface.dart';
import 'edit_preview.dart';

/// Computes the "effective" (preview) elements and selection overlay during
/// an edit session.
///
/// In the preview/commit architecture:
/// - `state.domain.document.elements` remains persistent (unchanged) during
///   editing
/// - `state.application.interaction` carries the session transform
///   (dx/dy/angle/bounds...)
/// - rendering and hit-testing should use this effective preview
class EditPreviewEngine {
  EditPreviewEngine();

  EditPreview build({
    required DrawState state,
    required EditOperationRegistry editOperations,
  }) {
    final interaction = state.application.interaction;
    if (interaction is! EditingState) {
      return EditPreview.none;
    }

    final operation = editOperations.getOperation(interaction.operationId);
    if (operation == null) {
      return EditPreview.none;
    }

    try {
      return operation.buildPreview(
        state: state,
        context: interaction.context,
        transform: interaction.currentTransform,
      );
    } on EditError {
      return EditPreview.none;
    }
  }
}

File: packages/snow_draw_rust_engine/src/draw/edit/preview/edit_preview_engine.rs
#![allow(dead_code)]

use crate::draw::edit::edit_operations::DefaultEditOperationRegistry;
use crate::draw::edit::preview::edit_preview::EditPreview;
use crate::draw::models::draw_state::DrawState;
use crate::draw::models::interaction_state::InteractionState;

/// Computes the effective edit preview for rendering and hit-testing.
#[derive(Clone, Copy, Debug, Default)]
pub struct EditPreviewEngine;

impl EditPreviewEngine {
    pub const fn new() -> Self {
        Self
    }

    pub fn build(
        &self,
        state: &DrawState,
        edit_operations: &DefaultEditOperationRegistry,
    ) -> EditPreview {
        let InteractionState::Editing(editing) = &state.application.interaction else {
            return EditPreview::none();
        };

        let Some(operation) = edit_operations.get_operation(editing.operation_id) else {
            return EditPreview::none();
        };

        operation.build_preview(state, &editing.context, &editing.current_transform)
    }
}
